lib/python/dataset.cpp

```diff
@@ -186,24 +187,22 @@ Returned by :py:func:`DataArray.masks`)");
   Dict of data arrays with aligned dimensions.)");
   options.disable_function_signatures();
   dataset.def(
-      py::init([](const std::map<std::string, std::variant<Variable, DataArray>>
-                      &data,
-                  const std::map<std::string, Variable> &coords) {
+      py::init([](const py::object &data, const py::object &coords) {
         Dataset d;
-        for (auto &&[name, item] : data) {
-          auto visitor = [&d, n = name](auto &object) {
-            d.setData(std::string(n), std::move(object));
-          };
-          std::visit(visitor, item);
+        for (auto &&[name, item] : py::dict(data)) {
+          if (py::isinstance<DataArray>(item)) {
+            d.setData(name.cast<std::string>(), item.cast<DataArray &>());
+          } else {
+            d.setData(name.cast<std::string>(), item.cast<Variable &>());
+          }
         }
-        for (auto &&[dim, coord] : coords)
-          d.setCoord(Dim{dim}, coord);
```

Comment:
Not related to your change, but I think we should *first* add coords *then* data? Otherwise coords of items will be replaced, without checking for consistency?

Reply:
What consistency? If we did that, coords of item would replace the explicitly given ones. I think this is the wrong way around.

Reply:
I mean check for compatible coords. `setCoord` does a plain replace, but when you set a new data item that has coords they will get compared to the one already in the dataset.